python/ray/serve/controller.py

```diff
@@ -374,33 +378,35 @@ def deploy(
         # the only change was num_replicas, the start_time_ms is refreshed.
         # Is this the desired behaviour?
 
-        updating = self.deployment_state_manager.deploy(name, deployment_info)
+        updating = await self.deployment_state_manager.deploy(name, deployment_info)
 
         if route_prefix is not None:
             endpoint_info = EndpointInfo(route=route_prefix)
-            self.endpoint_state.update_endpoint(name, endpoint_info)
+            await self.endpoint_state.update_endpoint(name, endpoint_info)
         else:
-            self.endpoint_state.delete_endpoint(name)
+            await self.endpoint_state.delete_endpoint(name)
 
         return updating
 
-    def deploy_group(self, deployment_args_list: List[Dict]) -> List[bool]:
+    async def deploy_group(self, deployment_args_list: List[Dict]) -> List[bool]:
         """
         Takes in a list of dictionaries that contain keyword arguments for the
         controller's deploy() function. Calls deploy on all the argument
         dictionaries in the list. Effectively executes an atomic deploy on a
         group of deployments.
         """
 
-        return [self.deploy(**args) for args in deployment_args_list]
+        return await asyncio.gather(
+            *[self.deploy(**args) for args in deployment_args_list]
```

Comment:
Now that each `deploy` call here is asynchronous, is there a chance that the Serve controller calls `deploy_group` and then calls `run_control_loop` before `deploy_group` finishes executing?

If so, that could be a bit problematic. The synchronous `deploy_group` is used to atomically deploy a group of deployments. This is useful in case any deployment tries to access the Serve handle of another deployment from the group during initialization.

Could we somehow keep `deploy_group` and `deploy` synchronous (or at least provide an option that makes them synchronous)?

---

For extra context, a deployment's handle is only available once the controller executes `deploy` on deployment (specifically after [this line](https://github.com/ray-project/ray/blob/62143e8e18eadfa1894ef44f4aa97782fc223c06/python/ray/serve/controller.py#L391) finishes). However, the controller will initialize each deployment only inside `run_control_loop`. A synchronous `deploy_group` is useful since it forces the controller to execute `deploy` on a group of deployments before calling `run_control_loop` again, allowing any deployment in the group to access any other deployment's handle during initialization.

Reply:
My feeling is that it won't happen. Let me know if I understand it incorrectly:

- run_control_loop will only be called once in init stage
- when serve got updates it'll always go to run_control_loop
- run_control_loop will call deploy (autoscaler) and start to deploy
- your concern here is that:
  - the deploy is async and before it finished, it'll return and run_control_loop will continue to run (sleep and deploy)

If this is true, I don't think it's an issue since although it's async run, it still guarantee the ordering.

For this one: 
```
        return await asyncio.gather(
            *[self.deploy(**args) for args in deployment_args_list]
```

So if inside self.deploy it has some await, the next self.deploy might happen.

If we change it to
```
        return [ await self.deploy(**args) for args in deployment_args_list]
```

The ordering will be: unless the first one finished, the next wont start.


For run_control_loop case, it's like:

1. await autoscaler
2. something
3. ..
4. asyncio.sleep()
5. go to 1)

Here 2 won't be scheduled unless 1) is finished. It's not a detached task, the await will yield and until the coroutine finished, it'll resume from the place where it yield.

Reply:
Got it. Let me think of a way to better support this. I think the worse case is to make them a sync function in some way.